### ISSUE TITLE
Missing template features and syntaxes

### ISSUE THREAD

**Issue by NullVoxPopuli** [emberjs/rfcs#816]
tl;dr: this is a checklist / meta-issue

This is kind of a meta issue tracking everything the community (and myself) wants to add to templates, and without these features, it _feels like_ once you get to a big enough project, you feel pretty limited / have tied hands, if that makes sense. We can expand on the need, use cases, ergonomics, etc in separate issues.

In no particular order,

Features
- [ ] https://github.com/emberjs/rfcs/issues/805
- [ ] https://github.com/emberjs/rfcs/pull/593
- [ ] Named blocks as arguments 
  - https://github.com/emberjs/rfcs/issues/791
  - https://github.com/emberjs/rfcs/issues/735
  - https://github.com/emberjs/rfcs/pull/460#issuecomment-902961346
- [ ] Destructuring
  - https://github.com/emberjs/rfcs/issues/553
  - https://github.com/emberjs/rfcs/issues/602
- [ ] Context / component-tree scoped-services
  - https://github.com/emberjs/rfcs/issues/775
  - [ ] Error Boundaries
    - [ ] https://github.com/emberjs/rfcs/issues/513
    - [ ] https://github.com/emberjs/rfcs/issues/518
- [ ] https://github.com/emberjs/rfcs/issues/517
- [ ] Primitives
  - [ ] https://github.com/emberjs/rfcs/issues/623
  - [ ] https://github.com/emberjs/rfcs/issues/473
- [ ] https://github.com/emberjs/rfcs/pull/695
- [ ] Built-in helpers
  - https://github.com/emberjs/rfcs/issues/714
  - [ ] https://github.com/emberjs/rfcs/pull/562
  - [ ] https://github.com/emberjs/rfcs/pull/561
  - [ ] https://github.com/emberjs/rfcs/pull/560
  - [ ] https://github.com/emberjs/rfcs/pull/391

Consistency
- [ ] https://github.com/emberjs/rfcs/issues/803

**Comment by machty** [comment]
Thank you for compiling and maintaining this <3

**Comment by moomoo-dev** [comment]
Add the ability to do switch statements in user land

Example 
`{{switch this.case }}

  {{case "woof"}}
   <SomeWoof/>
  {{/case }}
  
  {{case "meow"}}
   <SomeMeow/>
  {{/case }}

{{/switch}}`

**Comment by johanrd** [comment]
also relevant here: 
- [ ] https://github.com/emberjs/rfcs/pull/621
     -  [x] https://github.com/handlebars-lang/handlebars-parser/pull/15

**Comment by bymetasoft** [comment]
Missing `@index`, `@first` and `@last` in `#each` as stated here: https://handlebarsjs.com/guide/builtin-helpers.html#each

**Comment by NullVoxPopuli** [comment]
`@index` and `@first` and `@last` won't be added, because:
- we already use `@things` for _arguments_
- we don't want to implement a variety of handlebars features because they cause ambiguity and make code less clear -- this has been this way for a while (for this reason), and as such, I like to say that we _don't use_ handlebars. (any feature that adds implicitly defined things is the core category of things we've removed) -- we've even forked the parser, and are iterating on improving it for just our usage
